In [5]:
import pandas as pd
import numpy as np

In [7]:
car_sale = pd.read_csv(r"C:\Users\shayj\Downloads\Online_Cars_Sale_Marketplace (2).csv")

In [9]:
car_sale = car_sale[car_sale['Price'] != 'Not Priced'] ## remove 'Not Priced' rows
car_sale['Price'] = car_sale['Price'].replace(r'[,$]', '', regex=True)  ## remove special characters $ and , 
car_sale['Price'] = car_sale['Price'].astype(int) 
car_sale['Used/New'] = car_sale['Used/New'].replace(r'.*Certified.*', 'Certified', regex=True) ## change all Certified to just Certified
car_sale[car_sale['Drivetrain'] != "-"] ## remove "-" in Drivetrain column
Min = car_sale['MinMPG']
Max = car_sale['MaxMPG']
car_sale['MaxMPG'] = np.maximum(Min, Max)
car_sale['MinMPG'] = np.minimum(Min, Max) ## Fix the min and max being in the wrong column
car_sale.drop(car_sale[(car_sale['MinMPG'] == 0)&(car_sale['MaxMPG'] == 0)].index, inplace = True) ## remove rows where min and max MPG are 0 
car_sale = car_sale.drop_duplicates() ## remove duplicates
car_sale            #Andrews clean data, will be using it to try and answer my question

Year     Make             Model Used/New   Price  ConsumerRating  \
0     2021  Porsche               911     Used  157993             4.9   
1     2010      BMW          128 128i     Used    8999             4.7   
2     2021      RAM     1500 Big Horn     Used   49997             4.8   
3     2021      RAM     1500 Big Horn     Used   42548             4.8   
4     2021      RAM     1500 Big Horn     Used   41623             4.8   
...    ...      ...               ...      ...     ...             ...   
9373  2004      BMW  Z4 2.5i Roadster     Used   17990             4.7   
9374  2019      BMW      Z4 sDrive30i     Used   46500             5.0   
9376  2014      BMW      Z4 sDrive35i     Used   34500             4.9   
9377  2015      BMW     Z4 sDrive35is     Used   49400             4.6   
9378  2003     Ford               ZX2     Used    6995             4.0   

      ConsumerReviews SellerType                                SellerName  \
0                   4     Dealer                             Porsche Marin   
1                  10     Dealer                        A & H Quality Cars   
2                  92     Dealer  Posner Park Chrysler Dodge Jeep Ram Fiat   
3                  92     Dealer                      Kia of Wesley Chapel   
4                  92     Dealer                    Greenway Nissan Venice   
...               ...        ...                                       ...   
9373               33     Dealer       Susquehanna Chrysler Dodge Jeep RAM   
9374                2     Dealer                           BMW of Sterling   
9376               23     Dealer                           Oxmoor Chrysler   
9377               14     Dealer                            Baron BMW MINI   
9378               14     Dealer                        Dutro Ford Company   

      SellerRating  ...         InteriorColor         Drivetrain MinMPG  \
0              4.3  ...            Slate Gray                  –     18   
1              4.1  ...                 Black   Rear-wheel Drive     18   
2              4.2  ...  Light Mountain Brown   Four-wheel Drive     19   
3              4.8  ...                 Black   Rear-wheel Drive     20   
4              4.4  ...                 Black   Four-wheel Drive     19   
...            ...  ...                   ...                ...    ...   
9373           4.8  ...                     –   Rear-wheel Drive     20   
9374           4.5  ...           Ivory White   Rear-wheel Drive     25   
9376           4.7  ...             Coral Red   Rear-wheel Drive     19   
9377           4.6  ...                 Black   Rear-wheel Drive     17   
9378           1.6  ...                     –  Front-wheel Drive     26   

     MaxMPG  FuelType                       Transmission  \
0        23  Gasoline  8-Speed Automatic with Auto-Shift   
1        28  Gasoline                          Automatic   
2        24  Gasoline                  8-Speed Automatic   
3        25  Gasoline                  8-Speed Automatic   
4        24  Gasoline                  8-Speed Automatic   
...     ...       ...                                ...   
9373     28  Gasoline                          Automatic   
9374     32  Gasoline                  8-Speed Automatic   
9376     26  Gasoline  7-Speed Automatic with Auto-Shift   
9377     24  Gasoline  7-Speed Automatic with Auto-Shift   
9378     33  Gasoline                     5-Speed Manual   

                                                 Engine                VIN  \
0                       3.0L H6 24V GDI DOHC Twin Turbo  WP0CB2A96MS248778   
1     3.0L DOHC 24-valve I6 engine -inc: double-VANO...  WBAUP9C54AVL89249   
2                                  5.7L V8 16V MPFI OHV  1C6SRFFT8MN589592   
3                                  5.7L V8 16V MPFI OHV  1C6RREFTXMN503675   
4                                  5.7L V8 16V MPFI OHV  1C6SRFBT9MN739392   
...                                                 ...                ...   
9373        

For my question, I will be doing number 5, which is,

Which car listings is the least expensive?
(sub question?) which car depreciates the most over time

In [13]:
car_sale_pricing = car_sale.sort_values(by='Price')
car_sale_pricing

Year           Make              Model Used/New   Price  ConsumerRating  \
4612  2006      Chevrolet          Impala LT     Used    2300             4.2   
3439  2009           Ford           Focus SE     Used    3995             4.4   
2704  2004        Hyundai        Elantra GLS     Used    3995             4.1   
2924  2003          Lexus   ES 300 Base (A5)     Used    3999             4.8   
147   2015       Chrysler        200 Limited     Used    4250             4.4   
...    ...            ...                ...      ...     ...             ...   
8226  2020    Lamborghini          Urus Base     Used  285000             4.7   
565   2022        Porsche            911 GT3     Used  309888             4.9   
1018  2013    Lamborghini  Aventador LP700-4     Used  349995             4.6   
3593  2017  Mercedes-Benz      G-Class G 550     Used  409999             3.7   
867   2020  Mercedes-Benz      AMG G 63 Base     Used  449996             4.1   

      ConsumerReviews SellerType                              SellerName  \
4612              113    Private                                 Graydon   
3439               70     Dealer          Rochester Wholesale Auto Sales   
2704               30     Dealer                            Lets Go Auto   
2924               50     Dealer                               MotorMall   
147               361     Dealer                   Texas Best Deal Autos   
...               ...        ...                                     ...   
8226                1     Dealer  Ferrari Maserati of Central New Jersey   
565                 1     Dealer                             Mark Motors   
1018                2     Dealer                              RP Exotics   
3593                2     Dealer                      Monaco Motor Group   
867                 3     Dealer                       McLaren Charlotte   

      SellerRating  ...   InteriorColor         Drivetrain MinMPG MaxMPG  \
4612           4.6  ...           Beige  Front-wheel Drive     19     27   
3439           4.0  ...  Charcoal Black  Front-wheel Drive      0     24   
2704           4.4  ...            Gray  Front-wheel Drive     24     32   
2924           3.1  ...               –  Front-wheel Drive     21     29   
147            1.8  ...           Black  Front-wheel Drive     23     36   
...            ...  ...             ...                ...    ...    ...   
8226           4.6  ...        Nero Ade    All-wheel Drive     12     17   
565            5.0  ...           Black   Rear-wheel Drive     15     18   
1018           5.0  ...               –    All-wheel Drive     10     16   
3593           5.0  ...           Black    All-wheel Drive     13     14   
867            4.4  ...     Classic Red    All-wheel Drive     13     15   

           FuelType                       Transmission  \
4612  Gasoline Fuel                                A/T   
3439       Gasoline                             Manual   
2704       Gasoline                  4-Speed Automatic   
2924       Gasoline                  5-Speed Automatic   
147        Gasoline                  9-Speed Automatic   
...             ...                                ...   
8226       Gasoline                  8-Speed Automatic   
565        Gasoline  7-Speed Automatic with Auto-Shift   
1018       Gasoline  7-Speed Automatic with Auto-Shift   
3593       Gasoline                  7-Speed Automatic   
867        Gasoline                  9-Speed Automatic   

                                             Engine                VIN  \
4612  240.0HP 3.9L V6 Cylinder Engine Gasoline Fuel  2G1WC581469316469   
3439           2.0L DOHC 16-valve I4 Duratec engine  1FAHP32N19W189931   
2704                          2.0L I4 16V MPFI DOHC  KMHDN46D64U846583   
2924                          3.0L V6 24V MPFI DOHC  JTHBF30G136018786   
147                           2.4L I4 16V MPFI SOHC  1C3CCCAB0FN698261   
...                                             ...               

In [8]:
car_sale_pricing.head()

Year       Make             Model Used/New  Price  ConsumerRating  \
4612  2006  Chevrolet         Impala LT     Used   2300             4.2   
3439  2009       Ford          Focus SE     Used   3995             4.4   
2704  2004    Hyundai       Elantra GLS     Used   3995             4.1   
2924  2003      Lexus  ES 300 Base (A5)     Used   3999             4.8   
147   2015   Chrysler       200 Limited     Used   4250             4.4   

      ConsumerReviews SellerType                      SellerName  \
4612              113    Private                         Graydon   
3439               70     Dealer  Rochester Wholesale Auto Sales   
2704               30     Dealer                    Lets Go Auto   
2924               50     Dealer                       MotorMall   
147               361     Dealer           Texas Best Deal Autos   

      SellerRating  ...   InteriorColor         Drivetrain MinMPG MaxMPG  \
4612           4.6  ...           Beige  Front-wheel Drive     19     27   
3439           4.0  ...  Charcoal Black  Front-wheel Drive      0     24   
2704           4.4  ...            Gray  Front-wheel Drive     24     32   
2924           3.1  ...               –  Front-wheel Drive     21     29   
147            1.8  ...           Black  Front-wheel Drive     23     36   

           FuelType       Transmission  \
4612  Gasoline Fuel                A/T   
3439       Gasoline             Manual   
2704       Gasoline  4-Speed Automatic   
2924       Gasoline  5-Speed Automatic   
147        Gasoline  9-Speed Automatic   

                                             Engine                VIN  \
4612  240.0HP 3.9L V6 Cylinder Engine Gasoline Fuel  2G1WC581469316469   
3439           2.0L DOHC 16-valve I4 Duratec engine  1FAHP32N19W189931   
2704                          2.0L I4 16V MPFI DOHC  KMHDN46D64U846583   
2924                          3.0L V6 24V MPFI DOHC  JTHBF30G136018786   
147                           2.4L I4 16V MPFI SOHC  1C3CCCAB0FN698261   

       Stock#  Mileage  
4612        –   121000  
3439  09FOCUS   120000  
2704    11501   228692  
2924    4508R    77648  
147    698261   119417  

[5 rows x 32 columns]

In [24]:
car_sale_pricing.head(5)

Year       Make             Model Used/New  Price  ConsumerRating  \
4612  2006  Chevrolet         Impala LT     Used   2300             4.2   
3439  2009       Ford          Focus SE     Used   3995             4.4   
2704  2004    Hyundai       Elantra GLS     Used   3995             4.1   
2924  2003      Lexus  ES 300 Base (A5)     Used   3999             4.8   
147   2015   Chrysler       200 Limited     Used   4250             4.4   

      ConsumerReviews SellerType                      SellerName  \
4612              113    Private                         Graydon   
3439               70     Dealer  Rochester Wholesale Auto Sales   
2704               30     Dealer                    Lets Go Auto   
2924               50     Dealer                       MotorMall   
147               361     Dealer           Texas Best Deal Autos   

      SellerRating  ...   InteriorColor         Drivetrain MinMPG MaxMPG  \
4612           4.6  ...           Beige  Front-wheel Drive     19     27   
3439           4.0  ...  Charcoal Black  Front-wheel Drive      0     24   
2704           4.4  ...            Gray  Front-wheel Drive     24     32   
2924           3.1  ...               –  Front-wheel Drive     21     29   
147            1.8  ...           Black  Front-wheel Drive     23     36   

           FuelType       Transmission  \
4612  Gasoline Fuel                A/T   
3439       Gasoline             Manual   
2704       Gasoline  4-Speed Automatic   
2924       Gasoline  5-Speed Automatic   
147        Gasoline  9-Speed Automatic   

                                             Engine                VIN  \
4612  240.0HP 3.9L V6 Cylinder Engine Gasoline Fuel  2G1WC581469316469   
3439           2.0L DOHC 16-valve I4 Duratec engine  1FAHP32N19W189931   
2704                          2.0L I4 16V MPFI DOHC  KMHDN46D64U846583   
2924                          3.0L V6 24V MPFI DOHC  JTHBF30G136018786   
147                           2.4L I4 16V MPFI SOHC  1C3CCCAB0FN698261   

       Stock#  Mileage  
4612        –   121000  
3439  09FOCUS   120000  
2704    11501   228692  
2924    4508R    77648  
147    698261   119417  

[5 rows x 32 columns]

This code shows the 5 cheapest cars. Some observations are that they are all used cars and also have lots of miles on them. They a lot in common; most of them are sold by a dealer, same fuel type, drivetrain, MinMPG and MaxMPG


In [64]:
cheapest_cars = car_sale_pricing.head(5)
cheapest_cars

Year       Make             Model Used/New  Price  ConsumerRating  \
4612  2006  Chevrolet         Impala LT     Used   2300             4.2   
3439  2009       Ford          Focus SE     Used   3995             4.4   
2704  2004    Hyundai       Elantra GLS     Used   3995             4.1   
2924  2003      Lexus  ES 300 Base (A5)     Used   3999             4.8   
147   2015   Chrysler       200 Limited     Used   4250             4.4   

      ConsumerReviews SellerType                      SellerName  \
4612              113    Private                         Graydon   
3439               70     Dealer  Rochester Wholesale Auto Sales   
2704               30     Dealer                    Lets Go Auto   
2924               50     Dealer                       MotorMall   
147               361     Dealer           Texas Best Deal Autos   

      SellerRating  ...   InteriorColor         Drivetrain MinMPG MaxMPG  \
4612           4.6  ...           Beige  Front-wheel Drive     19     27   
3439           4.0  ...  Charcoal Black  Front-wheel Drive      0     24   
2704           4.4  ...            Gray  Front-wheel Drive     24     32   
2924           3.1  ...               –  Front-wheel Drive     21     29   
147            1.8  ...           Black  Front-wheel Drive     23     36   

           FuelType       Transmission  \
4612  Gasoline Fuel                A/T   
3439       Gasoline             Manual   
2704       Gasoline  4-Speed Automatic   
2924       Gasoline  5-Speed Automatic   
147        Gasoline  9-Speed Automatic   

                                             Engine                VIN  \
4612  240.0HP 3.9L V6 Cylinder Engine Gasoline Fuel  2G1WC581469316469   
3439           2.0L DOHC 16-valve I4 Duratec engine  1FAHP32N19W189931   
2704                          2.0L I4 16V MPFI DOHC  KMHDN46D64U846583   
2924                          3.0L V6 24V MPFI DOHC  JTHBF30G136018786   
147                           2.4L I4 16V MPFI SOHC  1C3CCCAB0FN698261   

       Stock#  Mileage  
4612        –   121000  
3439  09FOCUS   120000  
2704    11501   228692  
2924    4508R    77648  
147    698261   119417  

[5 rows x 32 columns]

In [56]:
car_sale.columns.tolist()


['Year',
 'Make',
 'Model',
 'Used/New',
 'Price',
 'ConsumerRating',
 'ConsumerReviews',
 'SellerType',
 'SellerName',
 'SellerRating',
 'SellerReviews',
 'StreetName',
 'State',
 'Zipcode',
 'DealType',
 'ComfortRating',
 'InteriorDesignRating',
 'PerformanceRating',
 'ValueForMoneyRating',
 'ExteriorStylingRating',
 'ReliabilityRating',
 'ExteriorColor',
 'InteriorColor',
 'Drivetrain',
 'MinMPG',
 'MaxMPG',
 'FuelType',
 'Transmission',
 'Engine',
 'VIN',
 'Stock#',
 'Mileage']

In [78]:
cheapest_cars[['Make', 'Model', 'Year', 'Price', 'Mileage', 'SellerType', 'SellerName', 'SellerRating']]

Make             Model  Year  Price  Mileage SellerType  \
4612  Chevrolet         Impala LT  2006   2300   121000    Private   
3439       Ford          Focus SE  2009   3995   120000     Dealer   
2704    Hyundai       Elantra GLS  2004   3995   228692     Dealer   
2924      Lexus  ES 300 Base (A5)  2003   3999    77648     Dealer   
147    Chrysler       200 Limited  2015   4250   119417     Dealer   

                          SellerName  SellerRating  
4612                         Graydon           4.6  
3439  Rochester Wholesale Auto Sales           4.0  
2704                    Lets Go Auto           4.4  
2924                       MotorMall           3.1  
147            Texas Best Deal Autos           1.8

Additional observations:

When looking at the the cheapest cars, we can see they are lot older cars from the early 2000s
Lots of miles on the cars 
Seller rating averages around 4 stars with the exception of the 1.8 rating on the Chrysler
Most of them are sold by dealerships. The cheapest car in this data set is the Chevrolet Impala LT with $2300 dollars for the car it is ranked (4612)

In [17]:
cheapest_car = car_sale_pricing.head(1)
cheapest_car

Year       Make      Model Used/New  Price  ConsumerRating  \
4612  2006  Chevrolet  Impala LT     Used   2300             4.2   

      ConsumerReviews SellerType SellerName  SellerRating  ...  InteriorColor  \
4612              113    Private    Graydon           4.6  ...          Beige   

             Drivetrain MinMPG MaxMPG       FuelType  Transmission  \
4612  Front-wheel Drive     19     27  Gasoline Fuel           A/T   

                                             Engine                VIN  \
4612  240.0HP 3.9L V6 Cylinder Engine Gasoline Fuel  2G1WC581469316469   

      Stock#  Mileage  
4612       –   121000  

[1 rows x 32 columns]

In [87]:
car_sale

Year     Make             Model Used/New   Price  ConsumerRating  \
0     2021  Porsche               911     Used  157993             4.9   
1     2010      BMW          128 128i     Used    8999             4.7   
2     2021      RAM     1500 Big Horn     Used   49997             4.8   
3     2021      RAM     1500 Big Horn     Used   42548             4.8   
4     2021      RAM     1500 Big Horn     Used   41623             4.8   
...    ...      ...               ...      ...     ...             ...   
9373  2004      BMW  Z4 2.5i Roadster     Used   17990             4.7   
9374  2019      BMW      Z4 sDrive30i     Used   46500             5.0   
9376  2014      BMW      Z4 sDrive35i     Used   34500             4.9   
9377  2015      BMW     Z4 sDrive35is     Used   49400             4.6   
9378  2003     Ford               ZX2     Used    6995             4.0   

      ConsumerReviews SellerType                                SellerName  \
0                   4     Dealer                             Porsche Marin   
1                  10     Dealer                        A & H Quality Cars   
2                  92     Dealer  Posner Park Chrysler Dodge Jeep Ram Fiat   
3                  92     Dealer                      Kia of Wesley Chapel   
4                  92     Dealer                    Greenway Nissan Venice   
...               ...        ...                                       ...   
9373               33     Dealer       Susquehanna Chrysler Dodge Jeep RAM   
9374                2     Dealer                           BMW of Sterling   
9376               23     Dealer                           Oxmoor Chrysler   
9377               14     Dealer                            Baron BMW MINI   
9378               14     Dealer                        Dutro Ford Company   

      SellerRating  ...  MinMPG MaxMPG  FuelType  \
0              4.3  ...      18     23  Gasoline   
1              4.1  ...      18     28  Gasoline   
2              4.2  ...      19     24  Gasoline   
3              4.8  ...      20     25  Gasoline   
4              4.4  ...      19     24  Gasoline   
...            ...  ...     ...    ...       ...   
9373           4.8  ...      20     28  Gasoline   
9374           4.5  ...      25     32  Gasoline   
9376           4.7  ...      19     26  Gasoline   
9377           4.6  ...      17     24  Gasoline   
9378           1.6  ...      26     33  Gasoline   

                           Transmission  \
0     8-Speed Automatic with Auto-Shift   
1                             Automatic   
2                     8-Speed Automatic   
3                     8-Speed Automatic   
4                     8-Speed Automatic   
...                                 ...   
9373                          Automatic   
9374                  8-Speed Automatic   
9376  7-Speed Automatic with Auto-Shift   
9377  7-Speed Automatic with Auto-Shift   
9378                     5-Speed Manual   

                                                 Engine                VIN  \
0                       3.0L H6 24V GDI DOHC Twin Turbo  WP0CB2A96MS248778   
1     3.0L DOHC 24-valve I6 engine -inc: double-VANO...  WBAUP9C54AVL89249   
2                                  5.7L V8 16V MPFI OHV  1C6SRFFT8MN589592   
3                                  5.7L V8 16V MPFI OHV  1C6RREFTXMN503675   
4                                  5.7L V8 16V MPFI OHV  1C6SRFBT9MN739392   
...                                                 ...                ...   
9373                              2.5L I6 24V MPFI DOHC  4USBT33534LS50632   
9374                         2.0L I4 16V GDI DOHC Turbo  WBAHF3C57KWW34090   
9376                    3.0L I6 24V GDI DOHC Twin Turbo  WBALM7C56EE386318   
9377                    3.0L I6 24V GDI DOHC Twin Turbo  WBALM1C50FE634478   
9378                              2.0L I4 16V MPFI DOHC  3FAFP11343R142429   

          Stock#  Mileage  Age  depreciation_rate  
0     1PMS248778    11472    4           39498.25

In [27]:
car_sale['Age'] = 2025 - car_sale['Year']
#figures out the age of the cars using 2025 as a refrence year



In [83]:

car_sale["depreciation_rate"] = car_sale["Price"] / car_sale["Age"]
car_sale["depreciation_rate"]


0       39498.250000
1         599.933333
2       12499.250000
3       10637.000000
4       10405.750000
            ...     
9373      856.666667
9374     7750.000000
9376     3136.363636
9377     4940.000000
9378      317.954545
Name: depreciation_rate, Length: 8471, dtype: float64

In [85]:
car_sale["depreciation_rate"] = car_sale["depreciation_rate"].round(2)
car_sale["depreciation_rate"]


0       39498.25
1         599.93
2       12499.25
3       10637.00
4       10405.75
          ...   
9373      856.67
9374     7750.00
9376     3136.36
9377     4940.00
9378      317.95
Name: depreciation_rate, Length: 8471, dtype: float64